### PR TITLE
[catbuffer/parser]: add comparer attribute to allow structs to define complex sorting rules

### PR DESCRIPTION
**Pull request by Jaguar0625**
problem: MultisigAccountModificationTransaction modifications need to be sorted with custom rules
solution: add comparer attribute to attach sort rules to MultisigAccountModification

### CODE REVIEW

File: catbuffer/parser/docs/cats_dsl.md [hunk 1]
```diff
@@ -224,6 +224,7 @@ Structures support the following attributes:
 1. `size(x)`: indicates that the `x` field contains the full size of the (variable sized) structure.
 1. `initializes(x, Y)`: indicates that the `x` field should be initialized with the `Y` constant.
 1. `discriminator(x [, y]+)`: indicates that the (`x`, ...`y`) properties should be used as the discriminator when generating a factory (only has meaning for abstract structures).
+1. `comparer(x [!a] [, y [!b]])`: indicates that the (`x`, ...`y`) properties should be used for custom sorting. optional (`a`, ...` b`) transforms can be specified and applied prior to property comparison. currently, the only transform supported is `keccak_256` for backwards compatibility with NEM.
```
Comment (gimre-xymcity) [inline catbuffer/parser/docs/cats_dsl.md:227]: ripemd_keccak

File: catbuffer/parser/catparser/grammar/catbuffer.lark [hunk 2]
```diff
@@ -44,11 +44,14 @@ struct_attribute: "@" ( \
...
 	| STRUCT_ATTRIBUTE_NAME_TWO_PARAMS "(" PROPERTY_NAME "," CONST_PROPERTY_NAME ")" \
 	| STRUCT_ATTRIBUTE_NAME_MULTI_PARAM "(" PROPERTY_NAME ("," PROPERTY_NAME)* ")" \
+	| STRUCT_ATTRIBUTE_NAME_MULTI_PARAM_WITH_TRANSFORM "(" PROPERTY_NAME ["!" TRANSFORM_NAME] ("," PROPERTY_NAME ["!" TRANSFORM_NAME])* ")" \
 ) _NL
 STRUCT_ATTRIBUTE_NAME_ZERO_PARAMS.1: "is_aligned" | "is_size_implicit"
 STRUCT_ATTRIBUTE_NAME_SINGLE_PARAM.1: "size"
 STRUCT_ATTRIBUTE_NAME_TWO_PARAMS.1: "initializes"
 STRUCT_ATTRIBUTE_NAME_MULTI_PARAM.1: "discriminator"
+STRUCT_ATTRIBUTE_NAME_MULTI_PARAM_WITH_TRANSFORM.1: "comparer"
+TRANSFORM_NAME.1: "keccak_256"
```
Comment (gimre-xymcity) [inline catbuffer/parser/catparser/grammar/catbuffer.lark:54]: one more

File: catbuffer/parser/tests/test_ast.py [hunk 3]
```diff
@@ -493,6 +494,24 @@ def test_can_create_struct_with_attribute_discriminator(self):
...
+		self.assertEqual(['alpha', 'beta'], [field.name for field in model.fields])
+		self._assert_disposition(model)
+		self._assert_attributes(model, comparer=[('beta', None), ('alpha', 'ripemd_keccak_256')])
+		self.assertEqual({
+			'name': 'FooBar',
+			'type': 'struct',
+			'layout': [{'name': 'alpha', 'type': 'MyCustomType'}, {'name': 'beta', 'size': 2, 'type': 'byte', 'signedness': 'unsigned'}],
+			'comparer': [{'name': 'beta', 'transform': None}, {'name': 'alpha', 'transform': 'ripemd_keccak_256'}]
+		}, model.to_legacy_descriptor())
+		self.assertEqual('@comparer(beta, alpha!keccak_256)\nstruct FooBar  # 2 field(s)', str(model))
```
Comment (gimre-xymcity) [inline catbuffer/parser/tests/test_ast.py:513]: hmm, ripemd_keccak? (is this passing?)

**Review by gimre-xymcity — APPROVED** [review]

File: sdk/python/scripts/ci/lint.sh [hunk 4]
```diff
@@ -17,7 +17,7 @@ find . -name "nc" -prune -o -name "sc" -prune -o -type f -name "*.py" -print0 |
 # generated code with some suppressions
 find . -name "nc" -o -name "sc" -o -type f -name "*.py" -print0 | PYTHONPATH=. xargs -0 python3 -m pycodestyle \
 	--config="$(git rev-parse --show-toplevel)/linters/python/.pycodestyle" \
-	--max-line-length=210
+	--max-line-length=215
```
[the remark this answers is not in the document]
Comment (Jaguar0625) [reply sdk/python/scripts/ci/lint.sh:20]: generated code doesn't conform to normal 140 length 😅

File: sdk/javascript/src/nem/models.js [hunk 5]
```diff
@@ -2734,6 +2734,15 @@ class MultisigAccountModification {
 		this._cosignatoryPublicKeySize = 32; // reserved field
 	}
 
+	comparer() {
+		const { ripemdKeccak256 } = require('../utils/transforms'); // eslint-disable-line global-require
```
Comment (gimre-xymcity) [inline sdk/javascript/src/nem/models.js:2738]: hmm, doesn't eslint require two spaces prior to `//` ?

**Review by gimre-xymcity — APPROVED** [review]

File: sdk/javascript/generator/StructTypeFormatter.py [hunk 6]
```diff
@@ -129,6 +129,29 @@ def get_ctor_descriptor(self):
...
+		if any('ripemd_keccak_256' == transform for (_, transform) in self.struct.comparer):
+			body += 'const { ripemdKeccak256 } = require(\'../utils/transforms\'); // eslint-disable-line global-require\n\n'
+
+		body += 'return [\n'
+		for (property_name, transform) in self.struct.comparer:
+			body += '\t'
+			if not transform:
+				body += f'this.{lang_field_name(property_name)}'
+			else:
+				body += f'{lang_field_name(transform).replace("_", "")}(this.{lang_field_name(property_name)}.bytes)'
```
Comment (gimre-xymcity) [inline sdk/javascript/generator/StructTypeFormatter.py:146]: given `.bytes` I assume this handles only array-like types now, right?

File: sdk/javascript/test/utils/transforms_spec.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+const { hexToUint8 } = require('../../src/utils/converter');
+const { ripemdKeccak256 } = require('../../src/utils/transforms');
+const { expect } = require('chai');
+
+describe('transforms', () => {
+	it('can transform with ripemd keccak 256', () => {
+		// Arrange:
+		const payload = hexToUint8('BE0B4CF546B7B4F4BBFCFF9F574FDA527C07A53D3FC76F8BB7DB746F8E8E0A9F');
```
Comment (gimre-xymcity) [inline sdk/javascript/test/utils/transforms_spec.js:8]: would rather use one of existing pub keys:
`75D8BB873DA8F5CCA741435DE76A46AFC2840803EBF080E931195B048D77F88C` - from cpp tests
`D6C3845431236C5A5A907A9E45BD60DA0E12EFD350B970E7F58E3499E2E7A2F0` - from js and py tests

File: sdk/python/scripts/ci/lint.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (gimre-xymcity) [reply sdk/python/scripts/ci/lint.sh:20]: ah this is for generated, fine 😂

File: sdk/python/symbolchain/Transforms.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import hashlib
+
+import sha3
+
+
+def ripemd_keccak_256(payload):
+	"""Hashes payload with keccak 256 and then hashes the result with ripemd160."""
+
+	outer_hasher = hashlib.new('ripemd160')
+	outer_hasher.update(sha3.keccak_256(payload).digest())
```
Comment (gimre-xymcity) [inline sdk/python/symbolchain/Transforms.py:10]: erm, not sure how I missed it, but this ofc is wrong 😬 

we'd need to do more or less what address is doing:
 * keccak
 * ripemd
 * concat proper pieces

**Review by gimre-xymcity — CHANGES_REQUESTED** [review]
ripemd_keccak_256

File: sdk/javascript/src/nem/models.js (hunk 5 above)
Comment (Jaguar0625) [reply sdk/javascript/src/nem/models.js:2738]: no, only pylint requires that

File: sdk/javascript/generator/StructTypeFormatter.py (hunk 6 above)
Comment (Jaguar0625) [reply sdk/javascript/generator/StructTypeFormatter.py:146]: correct

File: sdk/javascript/test/utils/transforms_spec.js (hunk 7 above)
Comment (Jaguar0625) [reply sdk/javascript/test/utils/transforms_spec.js:8]: used key taken from examples

File: sdk/javascript/generator/StructTypeFormatter.py (hunk 6 above)
Comment (gimre-xymcity) [reply sdk/javascript/generator/StructTypeFormatter.py:146]: should that go in a `.md` (not sure, asking)

File: sdk/javascript/test/utils/transforms_spec.js (hunk 7 above)
Comment (gimre-xymcity) [reply sdk/javascript/test/utils/transforms_spec.js:8]: yeah, but that key is not used to derive address :D
the two above are

File: sdk/python/symbolchain/Transforms.py (hunk 8 above)
Comment (gimre-xymcity) [reply sdk/python/symbolchain/Transforms.py:10]: not enough ☕, I guess this is negligible given need to create rmd collision to achieve that

File: sdk/javascript/generator/StructTypeFormatter.py (hunk 6 above)
Comment (Jaguar0625) [reply sdk/javascript/generator/StructTypeFormatter.py:146]: i think we would add support for non-array, if needed, so would not add to md

File: sdk/javascript/test/utils/transforms_spec.js (hunk 7 above)
Comment (Jaguar0625) [reply sdk/javascript/test/utils/transforms_spec.js:8]: idk, that key is actually hashed for sorting, so i thought it was better pick

**Pull request merged**